server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -2338,7 +2358,7 @@ public void waitForOpsToComplete(long seqNo) throws InterruptedException {
      */
     protected final boolean hasBeenProcessedBefore(Operation op) {
         assert op.seqNo() != SequenceNumbers.UNASSIGNED_SEQ_NO : "operation is not assigned seq_no";
-        assert versionMap.assertKeyedLockHeldByCurrentThread(op.uid().bytes());
+        assert noOpKeyedLock.isHeldByCurrentThread(op.seqNo()) || versionMap.assertKeyedLockHeldByCurrentThread(op.uid().bytes());
```

Comment:
can we condition this on the op type?

Reply:
++

Reply:
I pushed https://github.com/elastic/elasticsearch/pull/34408/commits/1df9b22c0c303a95c1e5a95d0963633d7c50245e